disk/src/com/treode/disk/edit/DriveGroup.scala

```diff
@@ -221,43 +225,60 @@ private class DriveGroup (
       // - make the drive, and
       // - add it to the list of new attaches.
       for (a <- change.attaches) {
-        if (attached contains a.path)
-          throw new IllegalArgumentException (s"Already attached ${quote (a.path)}")
-        if (attaching contains a.path)
-          throw new IllegalArgumentException (s"Already attaching ${quote (a.path)}")
-        attaching += a.path
-        val file = files.open (a.path, READ, WRITE)
-        val drive = new Drive (file, a.geometry, false, dno, a.path)
-        dno += 1
-        newAttaches ::= drive
+        if (attached contains a.path) {
+          errors.add (AlreadyAttached (a.path))
+        } else if (attaching contains a.path) {
+          errors.add (AlreadyAttaching (a.path))
+        } else {
+          attaching += a.path
+          val file = files.open (a.path, READ, WRITE)
+          val drive = new Drive (file, a.geometry, false, dno, a.path)
+          dno += 1
+          newAttaches ::= drive
+        }
       }
 
-      // The paths that are already queued draining.
-      var draining = (for (d <- drains) yield d.path).toSet
-
-      // Drives we will queue to drain; we build this up below.
-      var newDrains = List.empty [Drive]
+      // If there are attachment errors, close all opened files and pass errors.
+      if (errors.hasErrors) {
```

Comment:
I don't think we need to check for errors at this point. I think, even if there are attach errors, we can continue to process the drains also looking for drain errors. We only need to check for errors once, in the occurrence below. Just be sure to move the file.close down to that clause.